examples/cloudml-churn-prediction/preprocessor/preprocessor/preprocess.py

```diff
@@ -0,0 +1,342 @@
+    start_date = _random_start_date(d1, d2)
+    end_date = _random_end_date(start_date)
+    active = True if end_date > d2 else False
+    element.update(start_date=start_date)
+    if not active:
+        element.update(end_date=end_date)
+    else:
+        element.update(end_date=None)
+    element.update(active=active)
+    return element
+
+
+def _map_to_class(element):
+    """Extract duration and class from source data.
+
+    This function is required for both fake and real data, unless fields can be
+    extracted directly for source data (i.e. BQ).
+
+    Returns:
+        element with following fields added:
+            duration: number of days between subscription start date and end
+                date, or current date if customer is still active
+            label: String class denoting interval that duration falls in between
+    """
+
+    class_ceilings = features.LABEL_CEILINGS
+    if not element['active']:
+        duration = element['end_date'] - element['start_date']
+    else:
+        duration = datetime.datetime.now() - element['start_date']
+    duration_days = duration.days
+    element.update(duration=duration_days)
+    for index in range(0, len(class_ceilings)):
+        if duration_days < class_ceilings[index]:
+            element.update(label=features.LABEL_VALUES[index])
+            return element
+    element.update(label=features.LABEL_VALUES[len(class_ceilings)-1])
+    return element
+
+
+def _combine_censorship_duration(element):
+    """Create labels for training using lifetime and censorship.
+
+    Transform users' duration and censorship indicator into a 2*n_intervals
+    array. For each row, the first half of the array indicates whether the user
+    survived the corresponding interval. Second half indicates whether the user
+    died in the corresponding interval. If the user is censored, the second
+    half is all zeros.
+
+    This function is required for both fake and real data.
+
+    Args:
+        element: Dictionary representing one sample
+    Returns:
+        element with following field added:
+            labelArray: 2D array where for each row (which corresponds to a
+                user) first half of the values are 1 if the individual survived
+                that interval, 0 if not. The second half of the values are 1
+                for time interval during which failure occured (if uncensored)
+                and 0 for other intervals.
+    """
+
+    num_intervals = len(features.LABEL_CEILINGS)
+    y = np.zeros((num_intervals * 2))
+    breaks = [0] + features.LABEL_CEILINGS
+    duration = element['duration']
+    y[0:num_intervals] = 1.0*(np.full(num_intervals, duration) >= breaks[1:])
+    if not element['active']:
+        if duration < breaks[-1]:
+            y[num_intervals + np.where(np.full(
+                num_intervals, duration) < breaks[1:])[0][0]] = 1.0
+    element.update(labelArray=y)
+    return element
+
+
+@beam.ptransform_fn
+def shuffle(p):
```

Comment:
@sabhyankar I'm seeing this pattern pop up in a lot of ML SCE's beam pipelines.
In order to get randomly ordered, and randomly partitioned data, they are forcing a shuffle and having some kind of train-test-eval split function.
While PCollections don't guarantee ordering, they also don't randomize the data order unnecessarily. 
If an input PCollection has some ordering to it (say by date) and you write it you'll see ordered output [[ie. this gist](https://gist.github.com/jaketf/7b88578ee232316e38b5095957850795) demonstrating that while it forms random partitions, when writing each of them order is maintained].
Do you know any better way to randomize order of the data w/o forcing an expensive shuffle between nodes w/ `GroupByKey` ?

Reply:
@klmilam don't consider this comment blocking. sameer is OOO and I'm just curious if there's a better pattern here.